File: PracticaParcialTS/Ejercicio4.ts
//Red social con relaciones entre usuarios, grupos y páginas
class Usuario {
    private relaciones: Relacion[] = [];

    constructor(private nombre: string) {}

    // Método para agregar una relación con otro usuario
    agregarRelacion(relacion: Relacion): void {
        this.relaciones.push(relacion);
    }

    // Método para obtener el nombre del usuario
    obtenerNombre(): string {
        return this.nombre;
    }

    // Método para mostrar todas las relaciones del usuario
    mostrarRelaciones(): void {
        this.relaciones.forEach(relacion => {
            console.log(`${this.nombre} está relacionado con ${relacion.mostrar()}`);
        });
    }
}

class Relacion {
    constructor(private tipo: string, private conectadoCon: Usuario) {}

    // Método para mostrar la relación con el otro usuario
    mostrar(): string {
        return `${this.conectadoCon.obtenerNombre()} (Tipo: ${this.tipo})`;
    }
}

// Ejemplo de uso
const usuario1 = new Usuario("Juan");
const usuario2 = new Usuario("Ana");

// Crear relaciones entre los usuarios
usuario1.agregarRelacion(new Relacion("Amistad", usuario2));

// Mostrar las relaciones del usuario1
usuario1.mostrarRelaciones();

File: PracticaDeEjerciciosParcial/Ejercicio4.py
#Red social con relaciones entre usuarios, grupos y páginas
class Usuario:
    def __init__(self, nombre):
        self.nombre = nombre
        self.relaciones = []

    def agregar_relacion(self, relacion):
        self.relaciones.append(relacion)

    def mostrar_relaciones(self):
        for relacion in self.relaciones:
            print(f"{self.nombre} está relacionado con {relacion.mostrar()}")

class Relacion:
    def __init__(self, tipo, conectado_con):
        self.tipo = tipo
        self.conectado_con = conectado_con

    def mostrar(self):
        return f"{self.conectado_con.nombre} (Tipo: {self.tipo})"

class Grupo:
    def __init__(self, nombre):
        self.nombre = nombre

class Pagina:
    def __init__(self, nombre):
        self.nombre = nombre

# Ejemplo de uso
usuario1 = Usuario("Juan")
usuario2 = Usuario("Ana")
grupo = Grupo("Amantes de la tecnología")
pagina = Pagina("Página oficial de Python")

usuario1.agregar_relacion(Relacion("Amistad", usuario2))
usuario1.agregar_relacion(Relacion("Miembro", grupo))
usuario1.agregar_relacion(Relacion("Seguidor", pagina))

usuario1.mostrar_relaciones()
